README.txt
Please see our on-line docs for detailed instructions, examples, support forums, etc.:

http://simplegeo.com/docs/clients-code-libraries/python


See also these other two libraries:


https://github.com/simplegeo/python-simplegeo-shared

http://pypi.python.org/pypi/simplegeo-shared


https://github.com/simplegeo/python-simplegeo-places

http://pypi.python.org/pypi/simplegeo-places


debian/changelog
python-simplegeo-context (2.3.73) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * increase our requirement on the version of python-simplegeo-shared
    to 2.4 Because it in turn depends on a new version of python-oauth2
    which handles non-ascii chars better.

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 22:17:34 +0000

python-simplegeo-context (2.3.72) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 22:05:00 +0000

python-simplegeo-context (2.3.71) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 21:37:59 +0000

python-simplegeo-context (2.3.70) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 19:54:41 +0000

python-simplegeo-context (2.3.69) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 18:41:26 +0000

python-simplegeo-context (2.3.68) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 17:43:52 +0000

python-simplegeo-context (2.3.67) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 24 Jan 2011 17:22:32 +0000

python-simplegeo-context (2.3.66) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 21 Jan 2011 19:45:17 +0000

python-simplegeo-context (2.3.65) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 21 Jan 2011 19:43:02 +0000

python-simplegeo-context (2.3.64) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * raise requirement on python-simplegeo-shared from >= 2.2.45 to >=
    2.3.60
  * add test of oauth signing

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 21 Jan 2011 19:41:15 +0000

python-simplegeo-context (2.3.63) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 21 Jan 2011 19:33:32 +0000

python-simplegeo-context (2.3.62) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * update unit test to reflect new more lenient validation checking in
    python-simplegeo-shared

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Fri, 21 Jan 2011 19:01:14 +0000

python-simplegeo-context (2.3.61) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 04 Jan 2011 20:02:44 +0000

python-simplegeo-context (2.3.60) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 04 Jan 2011 19:12:52 +0000

python-simplegeo-context (2.3.59) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Thu, 23 Dec 2010 17:31:48 +0000

python-simplegeo-context (2.3.58) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * a tiny bit more type-checking and testing

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Wed, 22 Dec 2010 20:33:53 +0000

python-simplegeo-context (2.3.57) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * add get-context-by-address, version 2.3

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Wed, 22 Dec 2010 20:11:39 +0000

python-simplegeo-context (2.2.56) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * add docs for new get_context_by_* methods

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Wed, 22 Dec 2010 14:53:45 +0000

python-simplegeo-context (2.2.55) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * add get_context_by_my_ip() feature

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Wed, 22 Dec 2010 14:43:35 +0000

python-simplegeo-context (2.2.54) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Wed, 22 Dec 2010 14:40:25 +0000

python-simplegeo-context (2.2.53) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Wed, 22 Dec 2010 14:39:06 +0000

python-simplegeo-context (2.2.52) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Wed, 22 Dec 2010 14:19:30 +0000

python-simplegeo-context (2.2.51) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * this doesn't depend on simplejson directly -- this depends on
    python-simplegeo-shared which depends on pyutil[jsonutil] which
    (currently) depends on simplejson
  * split version into separately-editable manual version number and
    automatic build number
  * raise dependency on python-simplegeo-shared to >= 2.1 (needed for
    is_valid_ip())
  * raise dependency on python-simplegeo-shared to >= 2.2.45, to make
    sure we have the fix for the test of the obscure corner case in
    ipaddr

  [ Zooko ]
  * fix test to use something that is *really* not an IP address when
    testing that the client rejects non-IP-addresses

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Wed, 22 Dec 2010 06:58:05 +0000

python-simplegeo-context (2.0.44) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Tue, 21 Dec 2010 22:15:54 +0000

python-simplegeo-context (2.0.43) lucid; urgency=low

  [ Ben Standefer ]
  * Fixed consumption tests for classifiers rearrangement.

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 20 Dec 2010 22:24:53 +0000

python-simplegeo-context (2.0.42) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 20 Dec 2010 18:23:20 +0000

python-simplegeo-context (2.0.41) lucid; urgency=low

  * UNRELEASED

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Mon, 20 Dec 2010 17:34:15 +0000

python-simplegeo-context (2.0.40) lucid; urgency=low

  [ Ben Standefer ]
  * Make consumption tests check for dupes, more accurate, way better,
    etc.

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Thu, 09 Dec 2010 00:39:32 +0000

python-simplegeo-context (2.0.39) lucid; urgency=low

  [ Zooko Ofsimplegeo ]
  * increase requirement on -shared to >= 2.0

  [ SimpleGeo Nerds ]

 -- SimpleGeo Nerds <nerds@simplegeo.com>  Wed, 08 Dec 2010 07:34:32 +0000


test/test_client.py
import unittest, urllib

from simplegeo.context import Client
from simplegeo.shared import APIError, DecodeError, Feature

from decimal import Decimal as D

import mock

MY_OAUTH_KEY = 'MY_OAUTH_KEY'
MY_OAUTH_SECRET = 'MY_SECRET_KEY'
TESTING_LAYER = 'TESTING_LAYER'

API_VERSION = '1.0'
API_HOST = 'api.simplegeo.com'
API_PORT = 80

# example: http://api.simplegeo.com/0.1/context/37.797476,-122.424082.json

class ClientTest(unittest.TestCase):
    def setUp(self):
        self.client = Client(MY_OAUTH_KEY, MY_OAUTH_SECRET, API_VERSION, API_HOST, API_PORT)
        self.query_lat = D('37.8016')
        self.query_lon = D('-122.4783')

    def _record(self):
        self.record_id += 1
        self.record_lat = (self.record_lat + 10) % 90
        self.record_lon = (self.record_lon + 10) % 180

        return Feature(
            layer=TESTING_LAYER,
            id=str(self.record_id),
            coordinates=(self.record_lat, self.record_lon)
        )

    def test_wrong_endpoint(self):
        self.assertRaises(Exception, self.client._endpoint, 'wrongwrong')

    def test_missing_argument(self):
        self.assertRaises(Exception, self.client._endpoint, 'context')

    def test_get_context(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        res = self.client.get_context(self.query_lat, self.query_lon)
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s,%s.json' % (API_VERSION, self.query_lat, self.query_lon))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')
        # the code under test is required to have json-decoded this before handing it back
        self.failUnless(isinstance(res, dict), (type(res), repr(res)))

    @mock.patch('oauth2.Request.make_timestamp')
    @mock.patch('oauth2.Request.make_nonce')
    def test_oauth(self, mock_make_nonce, mock_make_timestamp):
        mock_make_nonce.return_value = 5
        mock_make_timestamp.return_value = 6

        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        self.client.get_context(self.query_lat, self.query_lon)

        self.assertEqual(mockhttp.method_calls[0][2]['body'], None)
        self.assertEqual(mockhttp.method_calls[0][2]['headers']['Authorization'], 'OAuth realm="http://api.simplegeo.com", oauth_nonce="5", oauth_timestamp="6", oauth_consumer_key="MY_OAUTH_KEY", oauth_signature_method="HMAC-SHA1", oauth_version="1.0", oauth_signature="AObXFB%2BqjwC20j5guAprkiREfIg%3D"')

    def test_get_context_by_address(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        addr = '41 Decatur St, San Francisco, CA'
        self.client.get_context_by_address(addr)
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/address.json?address=%s' % (API_VERSION, urllib.quote_plus(addr)))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_by_my_ip(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        self.client.get_context_by_my_ip()
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/ip.json' % (API_VERSION,))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_by_ip(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        ipaddr = '192.0.32.10'
        self.client.get_context_by_ip(ipaddr=ipaddr)
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s.json' % (API_VERSION, ipaddr))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_by_ip_invalid(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        self.failUnlessRaises(AssertionError, self.client.get_context_by_ip, '40.1,127.999')

    def test_get_context_invalid(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        self.failUnlessRaises(AssertionError, self.client.get_context, -91, 100)
        self.failUnlessRaises(AssertionError, self.client.get_context, -11, 361)

    def test_get_context_no_body(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, None)
        self.client.http = mockhttp

        self.failUnlessRaises(DecodeError, self.client.get_context, self.query_lat, self.query_lon)
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s,%s.json' % (API_VERSION, self.query_lat, self.query_lon))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_bad_json(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY + 'some crap')
        self.client.http = mockhttp

        try:
            self.client.get_context(self.query_lat, self.query_lon)
        except DecodeError, e:
            self.failUnlessEqual(e.code,None,repr(e.code))
            self.failUnless("Could not decode" in e.msg, repr(e.msg))
            repr(e)

        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s,%s.json' % (API_VERSION, self.query_lat, self.query_lon))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_error(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '500', 'content-type': 'application/json', }, '{"message": "help my web server is confuzzled"}')
        self.client.http = mockhttp

        try:
            self.client.get_context(self.query_lat, self.query_lon)
        except APIError, e:
            self.failUnlessEqual(e.code, 500, repr(e.code))
            self.failUnlessEqual(e.msg, '{"message": "help my web server is confuzzled"}', (type(e.msg), repr(e.msg)))

        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s,%s.json' % (API_VERSION, self.query_lat, self.query_lon))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_APIError(self):
        e = APIError(500, 'whee', {'status': "500"})
        self.failUnlessEqual(e.code, 500)
        self.failUnlessEqual(e.msg, 'whee')
        repr(e)
        str(e)


EXAMPLE_BODY="""
{
   "weather": {
    "message" : "'NoneType' object has no attribute 'properties'",
    "code" : 400
    },
   "features": [
    {
     "name" : "06075013000",
     "type" : "Census Tract",
     "bounds": [
      -122.437326,
      37.795016,
      -122.42360099999999,
      37.799485
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Census_Tract%3A06075013000%3A9q8zn0.json"
     },
     {
     "name" : "94123",
     "type" : "Postal",
     "bounds": [
      -122.452966,
      37.792787,
      -122.42360099999999,
      37.810798999999996
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Postal%3A94123%3A9q8zjc.json"
     },
     {
     "name" : "San Francisco",
     "type" : "County",
     "bounds": [
      -123.173825,
      37.639829999999996,
      -122.28178,
      37.929823999999996
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/County%3ASan_Francisco%3A9q8yvv.json"
     },
     {
     "name" : "San Francisco",
     "type" : "City",
     "bounds": [
      -123.173825,
      37.639829999999996,
      -122.28178,
      37.929823999999996
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/City%3ASan_Francisco%3A9q8yvv.json"
     },
     {
     "name" : "Congressional District 8",
     "type" : "Congressional District",
     "bounds": [
      -122.612285,
      37.708131,
      -122.28178,
      37.929823999999996
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Congressional_District%3ACongressional_Di%3A9q8yyn.json"
     },
     {
     "name" : "United States of America",
     "type" : "Country",
     "bounds": [
      -179.14247147726383,
      18.930137634111077,
      179.78114994357418,
      71.41217966730892
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Country%3AUnited_States_of%3A9z12zg.json"
     },
     {
     "name" : "Pacific Heights",
     "type" : "Neighborhood",
     "bounds": [
      -122.446782,
      37.787529,
      -122.422182,
      37.797728
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Neighborhood%3APacific_Heights%3A9q8yvz.json"
     },
     {
     "name" : "San Francisco1",
     "type" : "Urban Area",
     "bounds": [
      -122.51666666668193,
      37.19166666662851,
      -121.73333333334497,
      38.04166666664091
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Urban_Area%3ASan_Francisco1%3A9q9jsg.json"
     },
     {
     "name" : "California",
     "type" : "Province",
     "bounds": [
      -124.48200299999999,
      32.528832,
      -114.131211,
      42.009516999999995
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Province%3ACA%3A9qdguu.json"
     }
   ],
   "demographics": {
    "metro_score" : "10"
    }
   }
"""


simplegeo/context/test/test_client.py
import unittest, urllib

from simplegeo.context import Client
from simplegeo.shared import APIError, DecodeError, Feature

from decimal import Decimal as D

import mock

MY_OAUTH_KEY = 'MY_OAUTH_KEY'
MY_OAUTH_SECRET = 'MY_SECRET_KEY'
TESTING_LAYER = 'TESTING_LAYER'

API_VERSION = '1.0'
API_HOST = 'api.simplegeo.com'
API_PORT = 80

# example: http://api.simplegeo.com/0.1/context/37.797476,-122.424082.json

class ClientTest(unittest.TestCase):
    def setUp(self):
        self.client = Client(MY_OAUTH_KEY, MY_OAUTH_SECRET, API_VERSION, API_HOST, API_PORT)
        self.query_lat = D('37.8016')
        self.query_lon = D('-122.4783')

    def _record(self):
        self.record_id += 1
        self.record_lat = (self.record_lat + 10) % 90
        self.record_lon = (self.record_lon + 10) % 180

        return Feature(
            layer=TESTING_LAYER,
            id=str(self.record_id),
            coordinates=(self.record_lat, self.record_lon)
        )

    def test_wrong_endpoint(self):
        self.assertRaises(Exception, self.client._endpoint, 'wrongwrong')

    def test_missing_argument(self):
        self.assertRaises(Exception, self.client._endpoint, 'context')

    def test_get_context(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        res = self.client.get_context(self.query_lat, self.query_lon)
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s,%s.json' % (API_VERSION, self.query_lat, self.query_lon))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')
        # the code under test is required to have json-decoded this before handing it back
        self.failUnless(isinstance(res, dict), (type(res), repr(res)))

    def test_get_context_by_address(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        addr = '41 Decatur St, San Francisco, CA'
        self.client.get_context_by_address(addr)
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/address.json?address=%s' % (API_VERSION, urllib.quote_plus(addr)))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_by_my_ip(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        self.client.get_context_by_my_ip()
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/ip.json' % (API_VERSION,))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_by_ip(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        ipaddr = '192.0.32.10'
        self.client.get_context_by_ip(ipaddr=ipaddr)
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s.json' % (API_VERSION, ipaddr))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_by_ip_invalid(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        self.failUnlessRaises(AssertionError, self.client.get_context_by_ip, '40.1,127.999')

    def test_get_context_invalid(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY)
        self.client.http = mockhttp

        self.failUnlessRaises(AssertionError, self.client.get_context, -91, 100)
        self.failUnlessRaises(AssertionError, self.client.get_context, -11, 181)

    def test_get_context_no_body(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, None)
        self.client.http = mockhttp

        self.failUnlessRaises(DecodeError, self.client.get_context, self.query_lat, self.query_lon)
        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s,%s.json' % (API_VERSION, self.query_lat, self.query_lon))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_bad_json(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '200', 'content-type': 'application/json', }, EXAMPLE_BODY + 'some crap')
        self.client.http = mockhttp

        try:
            self.client.get_context(self.query_lat, self.query_lon)
        except DecodeError, e:
            self.failUnlessEqual(e.code,None,repr(e.code))
            self.failUnless("Could not decode" in e.msg, repr(e.msg))
            repr(e)

        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s,%s.json' % (API_VERSION, self.query_lat, self.query_lon))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_get_context_error(self):
        mockhttp = mock.Mock()
        mockhttp.request.return_value = ({'status': '500', 'content-type': 'application/json', }, '{"message": "help my web server is confuzzled"}')
        self.client.http = mockhttp

        try:
            self.client.get_context(self.query_lat, self.query_lon)
        except APIError, e:
            self.failUnlessEqual(e.code, 500, repr(e.code))
            self.failUnlessEqual(e.msg, '{"message": "help my web server is confuzzled"}', (type(e.msg), repr(e.msg)))

        self.assertEqual(mockhttp.method_calls[0][0], 'request')
        self.assertEqual(mockhttp.method_calls[0][1][0], 'http://api.simplegeo.com:80/%s/context/%s,%s.json' % (API_VERSION, self.query_lat, self.query_lon))
        self.assertEqual(mockhttp.method_calls[0][1][1], 'GET')

    def test_APIError(self):
        e = APIError(500, 'whee', {'status': "500"})
        self.failUnlessEqual(e.code, 500)
        self.failUnlessEqual(e.msg, 'whee')
        repr(e)
        str(e)


EXAMPLE_BODY="""
{
   "weather": {
    "message" : "'NoneType' object has no attribute 'properties'",
    "code" : 400
    },
   "features": [
    {
     "name" : "06075013000",
     "type" : "Census Tract",
     "bounds": [
      -122.437326,
      37.795016,
      -122.42360099999999,
      37.799485
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Census_Tract%3A06075013000%3A9q8zn0.json"
     },
     {
     "name" : "94123",
     "type" : "Postal",
     "bounds": [
      -122.452966,
      37.792787,
      -122.42360099999999,
      37.810798999999996
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Postal%3A94123%3A9q8zjc.json"
     },
     {
     "name" : "San Francisco",
     "type" : "County",
     "bounds": [
      -123.173825,
      37.639829999999996,
      -122.28178,
      37.929823999999996
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/County%3ASan_Francisco%3A9q8yvv.json"
     },
     {
     "name" : "San Francisco",
     "type" : "City",
     "bounds": [
      -123.173825,
      37.639829999999996,
      -122.28178,
      37.929823999999996
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/City%3ASan_Francisco%3A9q8yvv.json"
     },
     {
     "name" : "Congressional District 8",
     "type" : "Congressional District",
     "bounds": [
      -122.612285,
      37.708131,
      -122.28178,
      37.929823999999996
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Congressional_District%3ACongressional_Di%3A9q8yyn.json"
     },
     {
     "name" : "United States of America",
     "type" : "Country",
     "bounds": [
      -179.14247147726383,
      18.930137634111077,
      179.78114994357418,
      71.41217966730892
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Country%3AUnited_States_of%3A9z12zg.json"
     },
     {
     "name" : "Pacific Heights",
     "type" : "Neighborhood",
     "bounds": [
      -122.446782,
      37.787529,
      -122.422182,
      37.797728
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Neighborhood%3APacific_Heights%3A9q8yvz.json"
     },
     {
     "name" : "San Francisco1",
     "type" : "Urban Area",
     "bounds": [
      -122.51666666668193,
      37.19166666662851,
      -121.73333333334497,
      38.04166666664091
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Urban_Area%3ASan_Francisco1%3A9q9jsg.json"
     },
     {
     "name" : "California",
     "type" : "Province",
     "bounds": [
      -124.48200299999999,
      32.528832,
      -114.131211,
      42.009516999999995
     ],
     "href" : "http://api.simplegeo.com/0.1/boundary/Province%3ACA%3A9qdguu.json"
     }
   ],
   "demographics": {
    "metro_score" : "10"
    }
   }
"""


performance.py
#!/usr/bin/env python

import random
import time
import numpy

from simplegeo.context import Client
from simplegeo.shared import APIError

MY_OAUTH_KEY = ''
MY_OAUTH_SECRET = ''

API_VERSION = '1.0'
API_HOST = ''
API_PORT = 80

class PerformanceTest(object):
    def __init__(self, number_of_requests=10000):
        self.number_of_requests = number_of_requests
        self.client = Client(MY_OAUTH_KEY, MY_OAUTH_SECRET, API_VERSION, API_HOST, API_PORT)
        print self.client
        self.responses = []

    def run(self):
        requests_completed = 0
        requests_failed = 0

        for i in range(self.number_of_requests):
            try:
                print 'Trying request %s' % i
                timed_response = self._timed_response()
                requests_completed += 1
                self.responses.append(timed_response)
            except APIError:
                requests_failed += 1
                
        print self.responses
        print '%s requests completed, %s requests failed' % (requests_completed, requests_failed)

        times = [response['time_elapsed'] for response in self.responses]
        print '\n\nmin: %s max: %s avg: %s\n\n' % (min(times), max(times), (sum(times) / len(times)))
        bins = [i * 0.1 for i in range(20)] + [2.0] + [10.0]
        histogram = numpy.histogram(times, bins=bins)
        frequencies = [frequency for frequency in histogram[0]]
        bins = [bin for bin in histogram[1]]
        for (i, bin) in enumerate(bins):
            print '%ss' % bin
            try:
                print '\t' + str(frequencies[i]) + '\t' + '=' * frequencies[i]
            except:
                pass

    def _timed_response(self):
        (lat, lon) = self._random_lat_lon()
        print 'Timing client.get_context(%s, %s)' % (lat, lon)
        start_time = time.time()
        response = self.client.get_context(lat, lon)
        time_elapsed = time.time() - start_time
        print time_elapsed
        return {'lat': lat,
                'lon': lon,
                'response': response,
                'time_elapsed': time_elapsed}

    def _random_lat_lon(self):
        return (random.uniform(-90.0, 90.0), random.uniform(-180.0, 180.0))

def main():
    performance_test = PerformanceTest()
    performance_test.run()

if __name__ == '__main__':
    main()


__init__.py
API_VERSION = '1.0'

import urllib

# example: http://api.simplegeo.com/0.1/context/37.797476,-122.424082.json

from simplegeo.shared import json_decode
from simplegeo.shared import Client as SGClient, is_valid_ip, is_valid_lat, is_valid_lon
from pyutil.assertutil import precondition

class Client(SGClient):
    def __init__(self, key, secret, api_version=API_VERSION, host="api.simplegeo.com", port=80):
        SGClient.__init__(self, key, secret, api_version=api_version, host=host, port=port)

        self.endpoints['context'] = 'context/%(lat)s,%(lon)s.json'
        self.endpoints['context_by_ip'] = 'context/%(ip)s.json'
        self.endpoints['context_by_my_ip'] = 'context/ip.json'
        self.endpoints['context_by_address'] = 'context/address.json?address=%(address)s'

    def get_context(self, lat, lon):
        precondition(is_valid_lat(lat), lat)
        precondition(is_valid_lon(lon), lon)
        endpoint = self._endpoint('context', lat=lat, lon=lon)
        return json_decode(self._request(endpoint, "GET")[1])

    def get_context_by_ip(self, ipaddr):
        """ The server uses guesses the latitude and longitude from
        the ipaddr and then does the same thing as get_context(),
        using that guessed latitude and longitude."""
        precondition(is_valid_ip(ipaddr), ipaddr)
        endpoint = self._endpoint('context_by_ip', ip=ipaddr)
        return json_decode(self._request(endpoint, "GET")[1])

    def get_context_by_my_ip(self):
        """ The server gets the IP address from the HTTP connection
        (this may be the IP address of your device or of a firewall,
        NAT, or HTTP proxy device between you and the server), and
        then does the same thing as get_context_by_ip(), using that IP
        address."""
        endpoint = self._endpoint('context_by_my_ip')
        return json_decode(self._request(endpoint, "GET")[1])

    def get_context_by_address(self, address):
        """
        The server figures out the latitude and longitude from the
        street address and then does the same thing as get_context(),
        using that deduced latitude and longitude.
        """
        precondition(isinstance(address, basestring), address)
        endpoint = self._endpoint('context_by_address', address=urllib.quote_plus(address))
        return json_decode(self._request(endpoint, "GET")[1])

